### PR TITLE
data-source/aws_iam_policy_document: Add functionality to merge multiple policy documents

### PR DESCRIPTION
**Pull request by aborrello**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

### TL;DR

This PR introduces the two new attributes `source_json_list` and `override_json_list` into the `aws_iam_policy_document` data source. This enhancement allows for the layering and combining of IAM policy documents which is incredibly useful generating permission boundaries and reducing duplicated statements across Terraform definitions.

### Design

This PR models the behaviour described by @bflad in [this comment](https://github.com/terraform-providers/terraform-provider-aws/issues/5047#issuecomment-427118985) on issue #5047.

`source_json_list` provides a method to generate a base policy from multiple policy documents. Each non-empty SID across all statements (including statements defined in the `source_json`) must be unique, or a "Found duplicate sid" will be thrown.

`override_json_list` provides a method iteratively combine policy documents, performing an overwrite / merge on any duplicated SIDs.

This behaviour was achieved by utilising the `IAMPolicyDoc.Merge` method. The proposed conceptual flow for generating policy documents is as follows:

  1. Create an `IAMPolicyDoc` based on the provided `source_json`. If no `source_json` was defined, create a blank `IAMPolicyDoc`.
  1. *(NEW)* If the `source_json_list` attribute was specified, iterate over each policy and merge them. Before each individual policy is merged in, check that none of the non-blank SIDs have already been defined.
  1. Merge the statements defined in the current Terraform block's `statement` attributes.
  1. *(NEW)* If the `override_json_list` attribute was specified, iterate over each policy and merge them.
  1. Merge the `override_json` document if the attribute was specified.

### Tests

The following tests were added to verify this new functionality:
  - TestAccAWSDataSourceIAMPolicyDocument_sourceList
  - TestAccAWSDataSourceIAMPolicyDocument_sourceListConflicting
  - TestAccAWSDataSourceIAMPolicyDocument_overrideList

The outputs from these tests (and all other `TestAccAWSDataSourceIAMPolicyDocument`) can be found in the "Output from acceptance testing" section of this PR.

Closes the following issues:
  - #5047
  - #11942

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
data-source/aws_iam_policy_document: Add `source_json_list` attribute
data-source/aws_iam_policy_document: Add `override_json_list` attribute
```

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSDataSourceIAMPolicyDocument''

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSDataSourceIAMPolicyDocument -timeout 120m
=== RUN   TestAccAWSDataSourceIAMPolicyDocument_basic
=== PAUSE TestAccAWSDataSourceIAMPolicyDocument_basic
=== RUN   TestAccAWSDataSourceIAMPolicyDocument_source
=== PAUSE TestAccAWSDataSourceIAMPolicyDocument_source
=== RUN   TestAccAWSDataSourceIAMPolicyDocument_sourceList
=== PAUSE TestAccAWSDataSourceIAMPolicyDocument_sourceList
=== RUN   TestAccAWSDataSourceIAMPolicyDocument_sourceConflicting
=== PAUSE TestAccAWSDataSourceIAMPolicyDocument_sourceConflicting
=== RUN   TestAccAWSDataSourceIAMPolicyDocument_sourceListConflicting
=== PAUSE TestAccAWSDataSourceIAMPolicyDocument_sourceListConflicting
=== RUN   TestAccAWSDataSourceIAMPolicyDocument_override
=== PAUSE TestAccAWSDataSourceIAMPolicyDocument_override
=== RUN   TestAccAWSDataSourceIAMPolicyDocument_overrideList
=== PAUSE TestAccAWSDataSourceIAMPolicyDocument_overrideList
=== RUN   TestAccAWSDataSourceIAMPolicyDocument_noStatementMerge
=== PAUSE TestAccAWSDataSourceIAMPolicyDocument_noStatementMerge
=== RUN   TestAccAWSDataSourceIAMPolicyDocument_noStatementOverride
=== PAUSE TestAccAWSDataSourceIAMPolicyDocument_noStatementOverride
=== RUN   TestAccAWSDataSourceIAMPolicyDocument_duplicateSid
=== PAUSE TestAccAWSDataSourceIAMPolicyDocument_duplicateSid
=== RUN   TestAccAWSDataSourceIAMPolicyDocument_Statement_Principal_Identifiers_StringAndSlice
=== PAUSE TestAccAWSDataSourceIAMPolicyDocument_Statement_Principal_Identifiers_StringAndSlice
=== RUN   TestAccAWSDataSourceIAMPolicyDocument_Statement_Principal_Identifiers_MultiplePrincipals
=== PAUSE TestAccAWSDataSourceIAMPolicyDocument_Statement_Principal_Identifiers_MultiplePrincipals
=== RUN   TestAccAWSDataSourceIAMPolicyDocument_Version_20081017
=== PAUSE TestAccAWSDataSourceIAMPolicyDocument_Version_20081017
=== CONT  TestAccAWSDataSourceIAMPolicyDocument_basic
=== CONT  TestAccAWSDataSourceIAMPolicyDocument_noStatementMerge
=== CONT  TestAccAWSDataSourceIAMPolicyDocument_sourceListConflicting
=== CONT  TestAccAWSDataSourceIAMPolicyDocument_overrideList
=== CONT  TestAccAWSDataSourceIAMPolicyDocument_override
=== CONT  TestAccAWSDataSourceIAMPolicyDocument_Statement_Principal_Identifiers_MultiplePrincipals
=== CONT  TestAccAWSDataSourceIAMPolicyDocument_duplicateSid
=== CONT  TestAccAWSDataSourceIAMPolicyDocument_Version_20081017
=== CONT  TestAccAWSDataSourceIAMPolicyDocument_Statement_Principal_Identifiers_StringAndSlice
=== CONT  TestAccAWSDataSourceIAMPolicyDocument_noStatementOverride
=== CONT  TestAccAWSDataSourceIAMPolicyDocument_sourceList
=== CONT  TestAccAWSDataSourceIAMPolicyDocument_source
=== CONT  TestAccAWSDataSourceIAMPolicyDocument_sourceConflicting
--- PASS: TestAccAWSDataSourceIAMPolicyDocument_sourceListConflicting (11.10s)
--- PASS: TestAccAWSDataSourceIAMPolicyDocument_override (46.15s)
--- PASS: TestAccAWSDataSourceIAMPolicyDocument_Statement_Principal_Identifiers_MultiplePrincipals (46.34s)
--- PASS: TestAccAWSDataSourceIAMPolicyDocument_basic (46.37s)
--- PASS: TestAccAWSDataSourceIAMPolicyDocument_overrideList (46.48s)
--- PASS: TestAccAWSDataSourceIAMPolicyDocument_noStatementOverride (46.50s)
--- PASS: TestAccAWSDataSourceIAMPolicyDocument_sourceList (46.71s)
--- PASS: TestAccAWSDataSourceIAMPolicyDocument_sourceConflicting (46.71s)
--- PASS: TestAccAWSDataSourceIAMPolicyDocument_Statement_Principal_Identifiers_StringAndSlice (46.80s)
--- PASS: TestAccAWSDataSourceIAMPolicyDocument_noStatementMerge (46.80s)
--- PASS: TestAccAWSDataSourceIAMPolicyDocument_duplicateSid (50.75s)
--- PASS: TestAccAWSDataSourceIAMPolicyDocument_Version_20081017 (68.06s)
--- PASS: TestAccAWSDataSourceIAMPolicyDocument_source (72.62s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	72.661s
```

### CODE REVIEW

File: website/docs/d/iam_policy_document.html.markdown [hunk 1]
```diff
@@ -84,10 +84,21 @@ The following arguments are supported:
   current policy document.  Statements with non-blank `sid`s in the current
   policy document will overwrite statements with the same `sid` in the source
   json.  Statements without an `sid` cannot be overwritten.
+* `source_json_list` (Optional) - A list of IAM policy documents to import as
+  a base for the current policy document. Statements accross all policy documents
```
Comment (shadycuz) [inline website/docs/d/iam_policy_document.html.markdown]:
```suggestion
  a base for the current policy document. Statements across all policy documents
```

Just a minor typo

**Comment by shadycuz** [comment]
@aborrello Thanks for implementing these new features. It's going to make the `aws_iam_policy_document` so much more useful 🎉

File: website/docs/d/iam_policy_document.html.markdown [hunk 2]
```diff
@@ -84,10 +84,21 @@ The following arguments are supported:
...
+  defined in this list or in the `source_json` attribute must be unique. Statements
+  with non-blank `sid`s in the current policy document will overwrite statements
+  with the same `sid` from any of the source json documents.
 * `override_json` (Optional) - An IAM policy document to import and override the
   current policy document.  Statements with non-blank `sid`s in the override
-  document will overwrite statements with the same `sid` in the current document.
+  document will overwrite statements with the same `sid` in the current document,
+  including any defined by the `override_json_list` attribute.
   Statements without an `sid` cannot be overwritten.
+* `override_json_list` (Optional) - A list of IAM policy documents to import and
```
Comment (aborrello) [inline website/docs/d/iam_policy_document.html.markdown]: Open to suggestions around the documentation for this attribute. I found it quite hard to find a simple way to describe the iterative merge based on the ordering of the array...

File: aws/data_source_aws_iam_policy_document.go [hunk 3]
```diff
@@ -30,6 +30,11 @@ func dataSourceAwsIamPolicyDocument() *schema.Resource {
 				Type:     schema.TypeString,
 				Optional: true,
 			},
+			"override_json_list": {
```
Comment (aborrello) [inline aws/data_source_aws_iam_policy_document.go]: There was some discussion around `override_json_list` or `override_jsons`. I think the `_list` is cleaner, but completely open for discussion.

File: aws/data_source_aws_iam_policy_document.go [hunk 4]
```diff
@@ -112,6 +122,39 @@ func dataSourceAwsIamPolicyDocumentRead(d *schema.ResourceData, meta interface{}
...
+			sourceDoc := &IAMPolicyDoc{}
+			if err := json.Unmarshal([]byte(sourceJSON.(string)), sourceDoc); err != nil {
+				return err
+			}
+
+			// assure all statements in sourceDoc are unique before merging
+			for stmtIndex, stmt := range sourceDoc.Statements {
+				if stmt.Sid != "" {
+					if _, sidExists := sidMap[stmt.Sid]; sidExists {
+						return fmt.Errorf("Found duplicate sid (%s) in source_json_list (item %d; statement %d). Either remove the sid or ensure the sid is unique across all statements.", stmt.Sid, sourceJSONIndex, stmtIndex)
```
Comment (aborrello) [inline aws/data_source_aws_iam_policy_document.go]: This is slightly different to the normal error messages. I wanted to include the `json` index and `statement` index for developer ease, but am open to discuss different ways to convey this information.

**Comment by awoimbee** [comment]
This feature would be really helpful in a lot of scenarios and seem requested by a lot of people. (I'm really desperate for this feature 😅)
Any reason why this PR went stale ?

**Comment by YakDriver** [comment]
@aborrello Thank you for your work on this PR! I will be working on this before long. In order to expedite the process, I will likely make some changes. Make sure that you have checked the box "Allow edits from maintainers." (Don't worry though, you will receive all credit for your contribution and code!) Also, please coordinate with us before making any commits to this branch. Again, thank you for your help and we look forward to this popular addition to the AWS provider!

**Comment by YakDriver** [comment]
We have merged #12055 in to the Terraform AWS Provider. With this, [`aws_iam_policy_document`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/iam_policy_document) provides the ability to merge multiple policy documents. This is available now on the main branch and generally when version 3.28.0 is released (likely Feb. 11, 2021). If you have problems with the functionality or need further enhancements, please [open a new issue](https://github.com/hashicorp/terraform-provider-aws/issues/new/choose). Thanks for your interest in the AWS Provider! 🎉

**Comment by rednuht** [comment]
This cause this:
```
    statement {
        effect    = "Allow"
        actions   = [
            "s3:ListBucket"
        ]
        resources = [
           ....
        ]
        condition {
            test     = "ForAnyValue:StringLike"
            variable = "s3:prefix"
            values   = flatten([
                for path in [... some paths]:
                    ["${path}/*", path, ""]
            ])
        }
    }
```

To drop the `""` path. We use this to allow listing bucket objects on the "root" level and the paths specified in the condition.

**Pull request merged**